### ISSUE TITLE
[C++][feature request] extraction of the interface for multiple very simillar grammars

### ISSUE THREAD

**Issue by Nic30** [antlr/antlr4#2553]
Hello,
(I primary using C++ runtime but I beleive this feature would be useful for all runtimes)

I do have several grammars for different standard of the language.
For each standard I did generate parser.

The problem:
If I have a listener/visitor for some language version I have to manually go trough all rules in grammar and manually extract all functions for the same rules (but different type as they are for different parser).

This leads to ridiculous amount of work for grammars like Verilog/VHDL. In C++ I also have to rely on templatization heavily and still lots of code is just a type casting etc.

Proposed solution:
The goal is represent same rules in the grammar by same context type in  target language so they can be processed by same functions and most importantly it will be automatically checked that this rules are same and can be processed in this way.

The goal is achievable in multiple ways.
The way which seems to me as best now is:
1. Extract all context types for the same rules in to separate namespace.
2. Make virtual all context types for the rules which are used in same rules. And generate mapping virtual context type->context type for this parser for each parser.

**Comment by Thomasb81** [comment]
@Nic30 

I may not correctly capture the problem you are describing here. 

But in case you want to support several revision of a language, which basically have some rules addons from version to version, antlr4 has a feature call [Semantic Predicates](https://github.com/antlr/antlr4/blob/master/doc/predicates.md)

I have uses this feature here : 
[ `| {isSV2012() || isSV2017()}? file_nb `](https://github.com/Nic30/hdlConvertor/blob/f69214cdc9281a9834c96ceccefefaa34012197b/grammars/verilogPreprocParser.g4#L68) 
to support verilog and systemVerilog preprocessor in the same grammar.

Basically in the quoted line if `isSV2012()` or `isSV2017()` do not return true. The alternative rule `file_nb` cannot be select by the parser. The rule is not valid if we are not valid for verilog.

This lead to a single parser with several mode according to the flavor of the revision of the language you are trying to parse. This way no need to repeat code in a different listener/visitor of the various version of the grammar.

Thomas

**Comment by Nic30** [comment]
@Thomasb81, thank you I did not know about Sematic Predicates. My previous explanation of the problem is probably overcomplicated. I just wanted to share most of the code for multiple nearly same grammars.
